feat(new sink): add Apache Doris sink support

bingquanzhao opened the pull request:
## Summary

This PR introduces a new Apache Doris sink for Vector, enabling users to send log data directly to Apache Doris databases using the Stream Load API. The implementation includes:

- Complete Doris sink implementation with Stream Load API integration
- Comprehensive configuration options (endpoints, authentication, batching, custom headers)
- Full documentation generation using CUE
- Health check functionality with proper error handling
- Support for Doris-specific Stream Load parameters via custom HTTP headers

Apache Doris is a modern MPP analytical database that provides sub-second query response times on large datasets, making it ideal for real-time data warehouses and log analysis scenarios.

## Change Type
- [x] New feature
- [ ] Bug fix
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [ ] Yes
- [x] No

## How did you test this PR?

### Local Testing
1. **Unit Tests**: All unit tests pass with `cargo test`
2. **Configuration Validation**: Verified config parsing with `vector validate`
3. **Documentation Generation**: Successfully generated docs with `make generate-component-docs`
4. **CUE Validation**: All CUE files pass format and validation checks
5. **Changelog Validation**: Changelog fragment passes validation with `./scripts/check_changelog_fragments.sh`

### Test Configuration Used
```yaml
sources:
  demo:
    type: demo_logs
    format: json
    interval: 1

sinks:
  doris:
    type: doris
    inputs: ["demo"]
    
    # Target configuration
    endpoints: 
      - "http://doris-fe1:8030"
      - "http://doris-fe2:8030"
    database: "analytics_db"
    table: "user_events"
    
    # Authentication configuration
    auth:
      strategy: basic
      user: "admin"
      password: "admin123"
    
    # Batch configuration
    batch:
      max_events: 100000        # Maximum events per batch
      timeout_secs: 30          # Batch timeout in seconds
      max_bytes: 1073741824     # Maximum bytes per batch (1GB)
    
    # Custom HTTP headers for Doris Stream Load
    headers:
      format: "json"
      strip_outer_array: "false"
      read_json_by_line: "true"
    
    # Additional configuration
    label_prefix: "vector"
    log_request: true
    log_progress_interval: 10
    buffer_bound: 1
```

### Environment Setup
- Tested configuration validation against Vector's validation system
- Verified health check functionality (attempts connection to configured endpoints)
- All documentation generation and validation checks pass
- CUE v0.7.0 used for documentation generation

## Does this PR include user facing changes?

- [x] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Notes

### Implementation Details
- **Stream Load API**: Uses Doris's native Stream Load API for optimal performance and compatibility
- **Authentication**: Supports basic authentication with username/password
- **Batching**: Configurable batching with event count, byte size, and timeout limits
- **Custom Headers**: Support for Doris-specific Stream Load parameters via HTTP headers including:
  - `format`: Data format specification (json, csv, etc.)
  - `read_json_by_line`: JSON line-by-line reading mode
  - `strip_outer_array`: Array handling configuration
  - `columns`: Column mapping specification
- **Error Handling**: Comprehensive error handling with configurable retry logic
- **Health Checks**: Validates connectivity and basic authentication
- **Rate Limiting**: Built-in rate limiting and adaptive concurrency control

### Documentation
- Added complete CUE documentation for the sink configuration
- Generated reference documentation automatically using Vector's documentation system
- Updated service definitions and URL references
- All documentation validation checks pass (`CI=true make check-docs`)

### Dependencies
- No new external dependencies added
- Uses existing Vector HTTP client infrastructure
- Leverages standard Vector authentication, batching, and request frameworks
- Follows Vector's established patterns for sink implementation

### Code Quality
- All code formatted with `cargo fmt`
- Follows Vector's coding standards and patterns
- Proper error handling and logging throughout
- Comprehensive configuration validation

### Testing Strategy
- Configuration validation ensures all options are properly parsed
- Health check functionality verified through connection attempts
- Documentation generation confirms all metadata is correctly defined
- Follows Vector's established testing patterns for sinks

## References

- Apache Doris Stream Load Documentation: https://doris.apache.org/docs/data-operate/import/import-way/stream-load-manual
- Apache Doris Official Website: https://doris.apache.org
- Vector Sink Development Guidelines: https://vector.dev/docs/reference/configuration/sinks/

maycmlee commented on website/cue/reference/components/sinks/base/doris.cue:
```diff
@@ -0,0 +1,596 @@
+
+																Valid values are greater than or equal to `0`, and we expect reasonable values to range from `1.0` to `3.0`.
+
+																When calculating the past RTT average, we also compute a secondary “deviation” value that indicates how variable
```
```suggestion
																When calculating the past RTT average, we also compute a secondary "deviation" value that indicates how variable
```

maycmlee commented on website/cue/reference/components/sinks/base/doris.cue:
```diff
@@ -0,0 +1,596 @@
+															[arc]: https://vector.dev/docs/about/under-the-hood/networking/arc/
+															"""
+							none: """
+															A fixed concurrency of 1.
```
```suggestion
															A fixed concurrency of `1`.
```

maycmlee commented on website/cue/reference/components/sinks/base/doris.cue:
```diff
@@ -0,0 +1,596 @@
+				description: """
+					The amount of time to wait before attempting the first retry for a failed request.
+
+					After the first retry has failed, the fibonacci sequence is used to select future backoffs.
```
```suggestion
					After the first retry has failed, the Fibonacci sequence is used to select future backoffs.
```

maycmlee commented on website/cue/reference/components/sinks/base/doris.cue:
```diff
@@ -0,0 +1,596 @@
+						Full: """
+															Full jitter.
+
+															The random delay is anywhere from 0 up to the maximum current delay calculated by the backoff
```
```suggestion
															The random delay is anywhere from `0` up to the maximum current delay calculated by the backoff
```

maycmlee commented on website/cue/reference/components/sinks/doris.cue:
```diff
@@ -0,0 +1,197 @@
+		authentication: {
+			title: "Authentication"
+			body: """
+				Vector supports HTTP Basic Authentication for connecting to Doris. The
```
```suggestion
				Vector supports HTTP basic authentication for connecting to Doris. The
```

maycmlee commented on website/cue/reference/components/sinks/doris.cue:
```diff
@@ -0,0 +1,197 @@
+				Vector implements comprehensive error handling for Doris Stream Load operations:
+
+				- **Retries**: Failed requests are automatically retried based on the
+				  `max_retries` configuration. Set to -1 for unlimited retries.
```
```suggestion
				  `max_retries` configuration. Set to `-1` for unlimited retries.
```

bingquanzhao replied:
Thank you for your review. Your suggestions are very good. I will make the necessary revisions.

maycmlee reviewed (commented):
Some small suggestions

thomasqueirozb commented:
Still 12k lines added, find the offending commit and revert it or git checkout affected files from master

bingquanzhao commented:
Are there any other contents that need to be adjusted? Please help to check.

hanshuishi commented:
When can it be launched for use

pront commented:
Please merge / resolve conflicts with `origin/master`. The new `make fmt` should also update the formatting. In the meantime, we will review this PR soon.

pront commented:
Hello, I will review this tomorrow but due to the PR size, it's unlikely that we can merge it before the merge window closes.

pront commented on src/sinks/doris/config.rs:
```diff
@@ -0,0 +1,397 @@
+    /// The line delimiter for batch data.
+    #[configurable(metadata(docs::examples = "\\n"))]
+    #[serde(default = "default_line_delimiter")]
+    pub line_delimiter: String,
```
We probably need a framing config instead of this, see https://github.com/vectordotdev/vector/pull/23117/files#r2432978560

pront commented on src/sinks/doris/progress.rs:
```diff
@@ -0,0 +1,239 @@
+            };
+
+            // Output progress information using Vector's key-value format
+            info!(
```
These should be emitted as internal events. See `src/internal_events` for examples of how other sinks do this.

bingquanzhao commented on src/sinks/doris/config.rs:
```diff
@@ -0,0 +1,397 @@
+    /// Progress reporting interval in seconds.
+    /// Set to 0 to disable progress reporting.
+    #[serde(default = "default_log_progress_interval")]
+    pub log_progress_interval: u64,
```
log_progress can help users quickly troubleshoot writing problems during the writing process of Doris. At the same time, this function is consistent with various observable ecological components of Doris.

bingquanzhao replied:
If you strongly recommend using emit instead of this feature, we can discuss whether to keep this feature or not.

bingquanzhao commented:
> Hi @bingquanzhao, thank you for this PR. I did a review focusing on the config UX first.

Hi @pront ，your review comments were very helpful. I've made some code modifications; please continue to review them.

bingquanzhao commented:
Hi @pront ,

Thank you for your previous review! I've addressed all your feedback and also fixed the recent CI and test suite issues.
When you have time, I would really appreciate if you could take another look at the changes. Happy to make any additional improvements you suggest.

mozhu1024 commented:
Is this PR process still ongoing?

freejool commented:
@bingquanzhao Hi, Thanks for your work!

However, I got some warnings when consuming messages from kafka and loading into doris with your version (https://github.com/bingquanzhao/vector/commit/be278bb659da602ce140f151fedf8206f0e43299). Although messages are successfully written(unknown integrity).

> 2025-12-05T05:52:57.535661Z  WARN sink{component_kind="sink" component_id=doris_sink component_type=doris}: vector_buffers::buffer_usage_data: Buffer counter underflowed. Clamping value to `0`. current=2467 delta=2476

And here is my config:
```yaml
data_dir: /tmp/vector/data
sources:
  kafka_source:
    type: kafka
    bootstrap_servers: 100.88.1.4:9092
    group_id: vector_consumer111111111
    topics:
      - test_topic_1
    auto_offset_reset: earliest
    # 解析JSON消息
    decoding:
      codec: bytes
sinks:
  doris_sink:
    type: doris
    inputs:
      - kafka_source
    endpoints:
      - http://100.88.1.4:8030
    database: test
    table: "{{ message_key }}"
    auth:
      strategy: basic
      user: root
      password: "root@1234"
    encoding:
      codec: "text"
      only_fields: ["message"]
    # 启用批处理以提高性能
    batch:
      max_events: 1000
      timeout_secs: 1
    headers:
      format: "json"
      strip_outer_array: "false"
      read_json_by_line: "true"
    # 配置请求
    request:
      concurrency: 1
      rate_limit_duration_secs: 1
      rate_limit_num: 100
    # 配置重试
    acknowledgements:
      enabled: false
    log_request: true
```

Please ask me for any other information you need!

thomasqueirozb commented on src/internal_events/doris.rs:
```diff
@@ -0,0 +1,44 @@
+        warn!(
+            message = "Doris rows filtered during loading.",
+            filtered_rows = %self.filtered_rows,
+            internal_log_rate_limit = true
```
```suggestion
```

thomasqueirozb commented on website/cue/reference/components.cue:
```diff
@@ -222,7 +222,7 @@ components: {
 			}
 
-			exposes?: #FeaturesExpose
+			exposes?:                      #FeaturesExpose
```
```suggestion
			exposes?: #FeaturesExpose
```

This doesn't need to be changed

bingquanzhao replied:
It seems that I haven't made any changes here.

thomasqueirozb commented on src/sinks/doris/service.rs:
```diff
@@ -0,0 +1,151 @@
+                .send_stream_load(database, table, request.take_payload())
+                .await?;
+            let report_response = doris_response.clone();
+            let _ = service.reporter_run(report_response).await;
```
Why is the error silently ignored here? The future returns `Result<Self::Response, Self::Error>` and `Self::Error` has the same is the same error returned by `reporter_run`

thomasqueirozb commented on src/sinks/doris/client.rs:
```diff
@@ -0,0 +1,437 @@
+            message = "Building request.",
+            uri = %uri,
+            label = %label
+        );
```
```suggestion
        debug!(%uri, %label, "Building request.");
```
nit

thomasqueirozb commented on src/sinks/doris/client.rs:
```diff
@@ -0,0 +1,437 @@
+
+        // Extract response body
+        let (parts, body) = response.into_parts();
+        let body_bytes = hyper::body::to_bytes(body)
```
If you merge in master I'm 99% sure this will warn now after a recent PR of mine. See changes made in #24351 for a guide on how to fix the warning

thomasqueirozb commented on src/sinks/doris/health.rs:
```diff
@@ -0,0 +1,57 @@
+                        message = "Health check returned non-success status code, but not determining health state.",
+                        status_code = %status
+                    );
+                    None
```
Note to self: need to double check this behavior across other sources/sinks

thomasqueirozb commented on src/sinks/doris/common.rs:
```diff
@@ -0,0 +1,86 @@
+        let uri = format!("{}/_test", endpoint);
+        let uri = uri
+            .parse::<Uri>()
+            .context(InvalidHostSnafu { host: endpoint })?;
```
See previous comment about using `Uri`  instead of `String`

thomasqueirozb reviewed (commented):
Sorry for the delay! It was a lot of code 😁

bingquanzhao commented on src/sinks/doris/client.rs:
```diff
@@ -0,0 +1,437 @@
+        // Create and set headers
+        let mut headers = HashMap::new();
+        // Basic headers
+        headers.insert("Expect".to_string(), "100-continue".to_string());
```
Yes, this header is required by Doris Stream Load API. Without it, Doris returns an error: {"status":"FAILED","msg":"There is no 100-continue header"}.
The Expect: 100-continue mechanism allows the client to wait for server confirmation before sending the (potentially large) request body, avoiding unnecessary data transmission if the server would reject the request based on headers alone.
Reference: https://doris.apache.org/docs/data-operate/import/import-way/stream-load-manual

bingquanzhao commented:
> It actually looks like integration tests are failing when I run `./scripts/run-integration-tests int doris` due to a DB auth issue. I partially fixed the integration tests but didn't debug much further after I hit this issue

Thank you, I will try to address these issues as soon as possible.

freejool commented:
@bingquanzhao Hi, I added header group_commit: async_mode to conf, and got an error from doris "label and group_commit can't be set at the same time". Can you provide a switch to disable label generation? (more efficiency and less integrity)

bingquanzhao commented:
> @bingquanzhao Hi, I added header group_commit: async_mode to conf, and got an error from doris "label and group_commit can't be set at the same time". Can you provide a switch to disable label generation? (more efficiency and less integrity)

I will add a check. When group_commit is set, do not set the label.

freejool commented:
> > @bingquanzhao Hi, I added header group_commit: async_mode to conf, and got an error from doris "label and group_commit can't be set at the same time". Can you provide a switch to disable label generation? (more efficiency and less integrity)
> 
> The content you mentioned I have already pushed. You can give it a try.

Thanks! It works perfectly!

bingquanzhao commented:
Hi @thomasqueirozb , Do I still need to do anything else?

thomasqueirozb commented on src/sinks/doris/client.rs:
```diff
@@ -0,0 +1,518 @@
+            builder = builder.header("label", &label);
+            Some(label)
+        };
+        let _ = label; // Suppress unused variable warning
```
```suggestion
        if self.is_group_commit_enabled() {
            debug!(%uri, "Building request with group commit enabled (no label).");
        } else {
            let label = self.generate_label(database, table);
            debug!(%uri, %label, "Building request.");
            builder = builder.header("label", &label);
        };
```
